### ISSUE TITLE
how to make a nuscript standalone executable?

### ISSUE THREAD

**Issue by jakara** [nushell/nushell#8481]
### Question

like pyinstaller for python, is there a tool or method I can package a nuscript file into a standalone executable, that runs on OS without nushell installed?

**Comment by rgwood** [comment]
Nu does not have a way to do this right now. It's something we'd like to have eventually.

I'd lean toward taking the same approach as `deno compile`, where we embed the script inside a Nu executable: <https://github.com/denoland/deno/pull/8539> <https://github.com/denoland/deno/pull/9141>

```rust
/// This function will try to run this binary as a standalone binary
/// produced by `deno compile`. It determines if this is a stanalone
/// binary by checking for the magic trailer string `D3N0` at EOF-12.
/// After the magic trailer is a u64 pointer to the start of the JS
/// file embedded in the binary. This file is read, and run. If no
/// magic trailer is present, this function exits with Ok(()).
```
> `--target` allows to cross-compile binary to different target architectures by fetching appropriate binary from remote server on first run.

Related discussion on Discord: https://discordapp.com/channels/601130461678272522/615962413203718156/1077740195618758688

**Comment by jakara** [comment]
> 

no, I don't want to embed nushell in some seriously compiled(C/rust) programs, what I'm thinking is to make nu scripts more "portable",  so it can run on some bared OS, in bash maybe,  and it could work like some native tools, curl, eg.
I see nu language designed with a lot reliable, efficient features, such a waste to use as just a "shell", IMHO 😝

**Comment by rgwood** [comment]
I’m not sure I understand what you’re looking for. I agree that the Nu language is powerful and I’ve proposed one way that you could compile Nu scripts into standalone executables. Could you clarify what makes the  Pyinstaller approach different?

**Comment by jakara** [comment]
> I’m not sure I understand what you’re looking for. I agree that the Nu language is powerful and I’ve proposed one way that you could compile Nu scripts into standalone executables. Could you clarify what makes the Pyinstaller approach different?

Forget Pyinstaller, what `deno compile` did is exactly what I want from Nu

**Comment by salsa-dev** [comment]
@jakara having standalone script for execution could be achieved by using env in your script
```
#!/usr/bin/env nu
```
it is described in docs
https://www.nushell.sh/book/scripts.html#shebangs

this is probably the most you can achieve

**Comment by WindSoilder** [comment]
Closing the issue because the question is answered, currently you can use shebang to achieve something like this.

**Comment by XiNiHa** [comment]
This issue is about creating executables that can be run  **without having nushell installed,** so shebang can't be the answer. Reopen the issue please.

**Comment by fdncred** [comment]
Right now, the only way to do this is to build an executable with that script embedded similar to this https://github.com/sophiajt/nu_app or maybe something like this https://github.com/cablehead/how-to-embed-nu
